### PR TITLE
Background tasks

### PR DESCRIPTION
**Pull request by javierdelapuente**
Applicable spec: <link>
ISD181 - Background tasks in 12 factor apps

### Overview

This PR implement background services in 12 factor apps, using the `paas-app-charmer` library.

Services defined in `rockcraft.yaml` with their name ending in `-worker` will be injected the same environment variables as the main service and will these background services will run in all units.

Services defined in `rockcraft.yaml` with their name ending in `-scheduler` will be injected the same environment variables as the main service, but will only run in the unit ending in name 0. This pod will always exists if there at least one unit (juju does not set .spec.ordinals.start, see [here](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#ordinal-index) for reference.

The integration test for the PR is complex, as it uses Celery beat and Celery workers. The Celery beat will send tasks to the workers, with an argument that is the Celery beat hostname. The Celery workers will put in a Redis set the hostname of the beat scheduler (received as an argument) and its own hostname in another set. Checking the sets we can see how many schedulers and worker services are working in all the units in the application. 

As a simpler alternative, it could be possible to  directly check which services are started in each unit and if they contain environment variables.

Documentation will be handled in different PRs/posts.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [ ] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: paas_app_charmer/app.py [hunk 1]
```diff
@@ -192,6 +206,18 @@ def _app_layer(self) -> ops.pebble.LayerDict:
         services[self._workload_config.service_name]["override"] = "replace"
         services[self._workload_config.service_name]["environment"] = self.gen_environment()
 
+        for service_name, service in services.items():
+            # Add environment variables to all worker processes.
+            if service_name.endswith(WORKER_SUFFIX):
```
Comment (jdkandersson) [inline paas_app_charmer/app.py:211]: do you think we should be case-insensitive here?

File: paas_app_charmer/app.py [hunk 2]
```diff
@@ -192,6 +206,18 @@ def _app_layer(self) -> ops.pebble.LayerDict:
...
+                service["environment"] = self.gen_environment()
+            # For scheduler processes, add environment variables if
+            # the scheduler should run in the unit, disable it otherwise.
+            if service_name.endswith(SCHEDULER_SUFFIX):
```
Comment (jdkandersson) [inline paas_app_charmer/app.py:215]: Also here regarding case sensitivity

File: tests/unit/go/test_app.py [hunk 3]
```diff
@@ -94,6 +94,7 @@ def test_go_environment_vars(
         log_files=[],
         metrics_target=f"*:{framework_config.metrics_port}",
         metrics_path=framework_config.metrics_path,
+        unit_name="go/0",
```
Comment (jdkandersson) [inline tests/unit/go/test_app.py:97]: I saw this change for Flask, Django and go tests, but not for fast API?

Comment (javierdelapuente) [reply tests/unit/go/test_app.py:97]: I think the refactor for the App was done in Go, an that is why that test is there. 

I have reviewed the tests for FastAPI, and added one integration so the env variables for integrations are checked, that was missing.

File: paas_app_charmer/app.py (hunk 1 above)
Comment (javierdelapuente) [reply paas_app_charmer/app.py]: I have updated it so the worker name suffix can be case insensitive

**Review by jdkandersson — APPROVED** [review]

File: examples/flask/test_rock/app.py [hunk 4]
```diff
@@ -16,11 +18,86 @@
...
+        ),
+    )
+    celery_app.config_from_object(app.config["CELERY"])
+    return celery_app
+
+
 app = Flask(__name__)
 app.config.from_prefixed_env()
 
+broker_url = os.environ.get("REDIS_DB_CONNECT_STRING")
```
Comment (amandahla) [inline examples/flask/test_rock/app.py:58]: Minor: no need to validate the REDIS_DB_CONNECT_STRING?

**Review by amandahla — APPROVED** [review]
Nice!

File: examples/flask/test_rock/app.py (hunk 4 above)
Comment (javierdelapuente) [reply examples/flask/test_rock/app.py:58]: yeah, that's a good point, not validating the url will probably make celery work even if there is no broker (which is good and bad). I would leave like it is for now (and update it in another PR if it creates problems having celery crashing when there is no redis). 
Thanks for the tip!

**Pull request merged**